src/algebra/monoid_algebra.lean

```diff
@@ -373,6 +377,81 @@ end
 
 end misc_theorems
 
+/-! #### Non-unital, non-associative algebra structure -/
+section non_unital_non_assoc_algebra
+
+variables (k) [semiring k] [has_mul G]
+
+instance is_scalar_tower_self : is_scalar_tower k (monoid_algebra k G) (monoid_algebra k G) :=
+⟨λ t a b,
+begin
+  ext m,
+  simp only [mul_apply, finsupp.mul_sum, smul_apply, sum_smul_index, implies_true_iff,
+    eq_self_iff_true, smul_eq_mul, sum_zero, zero_mul, mul_assoc, mul_ite, mul_zero, if_t_t],
+end⟩
+
+/-- If the coefficients are commutative amongst themselves, the also commute with the algebra
+multiplication. -/
+instance smul_comm_class_self {k : Type u₁} [comm_semiring k] :
+  smul_comm_class k (monoid_algebra k G) (monoid_algebra k G) :=
+⟨λ t a b,
+begin
+  ext m,
+  simp only [mul_apply, finsupp.mul_sum, smul_apply, sum_smul_index, implies_true_iff,
+    eq_self_iff_true, smul_eq_mul, mul_ite, mul_zero, mul_left_comm, ite_eq_right_iff],
+end⟩
+
+instance smul_comm_class_symm_self {k : Type u₁} [comm_semiring k] :
+  smul_comm_class (monoid_algebra k G) k (monoid_algebra k G) :=
+smul_comm_class.symm k (monoid_algebra k G) (monoid_algebra k G)
```

Comment:
These two fall out as something like:
```lean
instance smul_comm_class_self [smul_comm_class R k k] :
  smul_comm_class R (monoid_algebra k G) (monoid_algebra k G) :=
⟨λ t a b,
begin
  ext m,
  simp [mul_apply, finsupp.sum, finset.smul_sum, smul_ite, mul_smul_comm, sum_smul_index'],
end⟩

instance smul_comm_class_symm_self [smul_comm_class k R k] :
  smul_comm_class (monoid_algebra k G) R (monoid_algebra k G) :=
⟨λ t a b,
begin
  haveI : smul_comm_class R k k := smul_comm_class.symm _ _ _,
  rw ←smul_comm,
end⟩
```
I'm sure the second one can be golfed.

Reply:
Done. Couldn't see an obvious golfing but did reduce the character count.